src/rgw/rgw_rest_pubsub.cc

```diff
@@ -714,7 +715,9 @@ void RGWPSCreateNotifOp::execute(optional_yield y) {
     // generate the internal topic. destination is stored here for the "push-only" case
     // when no subscription exists
     // ARN is cached to make the "GET" method faster
-    op_ret = ps.create_topic(this, unique_topic_name, topic_info.dest, topic_info.arn, topic_info.opaque_data, y);
+    op_ret = ps.create_topic(this, unique_topic_name, topic_info.dest,
```

Comment:
this is not a real topic that we create here. actually, this is the "notification" (or an internal/auto-generated topic).
so, the "owner" here may not eb the same as the owner of the topic (usually, here that would be the bucket owner).
we can keep it as is, and use the bucket owner here, or copy the value from `topic_info.user`?

either way, these topics are hidden the user via REST or radosgw-admin (see: https://github.com/ceph/ceph/blob/main/src/rgw/rgw_pubsub.cc#L401)

if we do want to present the owner of the "notification' (assuming we want to allow it to be different than the topic), we should add it to the JSON dump of `rgw_pubsub_topic_filter`. see: https://github.com/ceph/ceph/blob/main/src/rgw/rgw_pubsub.cc#L401

Reply:
i am kind a confused here, so from what i understand is that this code here generates an internal/topic while user subscribes bucket to a topic ( not sure why we create again an internal topic here)
so this means, that there could be a scenario where any s3 user can subscribe to any topic created by another s3 user, which technically does not make sense at all from S3 point. (S3 USER-A creates new topic <T1> and now S3 User-B creates a bucket "foo" and then subscribes notification for topic <T1> which was created by User-A) !
Ideally this should never be a scenario, but currently code does not enforce it ? So can we modify this code to enforce this here ?
 bucket owner has to be same as TOPIC owner ? thoughts ?

Reply:
> i am kind a confused here, so from what i understand is that this code here generates an internal/topic while user subscribes bucket to a topic ( not sure why we create again an internal topic here) 

this is indeed confusing, we use topics both for "topics" (=where to send the notifications to)  and for "notification configuration" (=the association of a bucket with a topic + filters). however, this is working, and refactoring this is lower priority

> so this means, that there could be a scenario where any s3 user can subscribe to any topic created by another s3 user, which technically does not make sense at all from S3 point. (S3 USER-A creates new topic and now S3 User-B creates a bucket "foo" and then subscribes notification for topic which was created by User-A) ! Ideally this should never be a scenario, but currently code does not enforce it ? So can we modify this code to enforce this here ? 

in AWS, topic creation is not part of the S3 API, so, I do think it does makes sense - an S3 user can subscribe to topic created by an SNS user.
we may need a better mechanism to control that.

> bucket owner has to be same as TOPIC owner ? thoughts ?

i don't think you want that. the topic creator is usually an admin of the system, and one topic may serve large number of buckets.
currently anyone can create a topic, and any bucket owner may use it to create notifications only on their bucket.
I think we should add topic management to the [user capability mechanism](https://docs.ceph.com/en/latest/radosgw/adminops/#add-a-user-capability), so that only specific users could create topics.
for the usecase of many buckets sending notification to the same broker, but with different ***kafka topics***, we may want to allow a different mechanism, maybe letting the "topic" name come from the notification configuration and not from the topic configuration.

Reply:
> in AWS, topic creation is not part of the S3 API, so, I do think it does makes sense - an S3 user can subscribe to topic created by an SNS user.
TOPIC creation is indeed part of S3 api ! you need S3 credentials to be able to create S3 topic. 

> i don't think you want that. the topic creator is usually an admin of the system, and one topic may serve large number of buckets.
And regarding the workflow that ceph cluster admin creates an topic and then the bucket owner subscribe to that topic, even that workflow is incorrect (applicable only for replication). In Conventional S3 notification, same S3 user will create a TOPIC and then subscribe its bucket to that topic, so the owner can create a workflow based on notification. For an SSL enabled topic that has user-id and password, topic definitely has to be created by bucket-owner, otherwise how will it even design a workflow based on notification, how will it be able to receive the notifications from the topic if it does not have access to kafka/aws/https topic server ?

Reply:
Currently the way we see the notification workflow is, individual S3 user will create its TOPICS and based on its workflow it will subscribe its bucket to respective topics which are created by himself, so the owner has the full control on topic it has created and can design the consuming of those topics.

Reply:
Agree as you pointed in the email, we can discuss offline more about the ownership .

Reply:
> This is the script we are using to create topic and it is totally controlled by s3 credentials

yes, in our implementation we only have S3 users, so we used the S3 user as a "hack" :-)
in AWS, these are unrelated services.
my point here is that there are AWS usecases that allow for different relationships between bucket owners and topic owners, and we want to allow this in our case as well.

Reply:
got it, so then we discuss all these scenario's offline ? eventually they all lead to ownership ?
for now we are good with this PR ?